pre_pull_base_image: add support for source_registry with auth

mlangsdorf opened the pull request:
fixes OSBS-6625 "Unable to use private quay.io org/repo directly"

With the introduction of quay support, there are private source registries that require auth.  Add support for retrieving their auth from REACTOR_CONFIG and passing it to the relevant functions.

Signed-off-by: Mark Langsdorf <mlangsdo@redhat.com>

rcerven commented on atomic_reactor/plugins/pre_reactor_config.py:
```diff
@@ -278,10 +278,13 @@ def get_source_registry(workflow, fallback=NO_FALLBACK):
         'insecure': source_registry.get('insecure', False)
     }
+    if source_registry.get('auth'):
+        regdict['secret'] = source_registry['auth']['cfg_path']
```
set 'secret' to None in case auth isn't defined

mlangsdorf replied:
Fixed.

MartinBasti commented:
This is from `store_metadata_in_osv3` plugin
```
2019-03-26 15:07:31,340 platform:- - atomic_reactor.plugins.store_metadata_in_osv3 - INFO - build id = docker-hello-world-multistage-51518-1
2019-03-26 15:07:31,340 - osbs.core - INFO - Using service account's auth token
2019-03-26 15:07:31,341 platform:- - atomic_reactor.util - DEBUG - query_registry: querying /v2/mbasti-osbs-box/fedora/manifests/latest, headers: {u'Accept': 'application/vnd.docker.distribution.manifest.v1+json'}
2019-03-26 15:07:31,578 platform:- - atomic_reactor.util - DEBUG - query_registry: [401] https://quay.io/v2/mbasti-osbs-box/fedora/manifests/latest
2019-03-26 15:07:31,578 platform:- - atomic_reactor.util - DEBUG - query_registry: [401] https://quay.io/v2/mbasti-osbs-box/fedora/manifests/latest
2019-03-26 15:07:31,578 platform:- - atomic_reactor.util - DEBUG - query_registry: response headers: {'content-length': '112', 'server': 'nginx/1.14.2', 'connection': 'keep-alive', 'docker-distribution-api-version': 'registry/2.0', 'date': 'Tue, 26 Mar 2019 15:07:32 GMT', 'content-type': 'application/json', 'www-authenticate': 'Bearer realm="https://quay.io/v2/auth",service="quay.io",scope="repository:mbasti-osbs-box/fedora:pull"'}
2019-03-26 15:07:31,578 platform:- - atomic_reactor.plugin - DEBUG - Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 257, in run
    plugin_response = plugin_instance.run()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugins/exit_store_metadata_in_osv3.py", line 216, in run
    base_image_id = self.workflow.builder.base_image_inspect['Id']
  File "/usr/lib/python2.7/site-packages/atomic_reactor/build.py", line 288, in base_image_inspect
    self.base_image_insecure)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1056, in get_inspect_for_image
    dockercfg_path=dockercfg_path)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1037, in get_all_manifests
    response, _ = get_manifest(image, registry_session, version)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 916, in get_manifest
    response = query_registry(registry_session, image, digest=None, version=version)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 853, in query_registry
    response.raise_for_status()
  File "/usr/lib/python2.7/site-packages/requests/models.py", line 834, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 401 Client Error: UNAUTHORIZED
```

MartinBasti commented:
Sorry wrong traceback, it was actually bump_release plugin:

```
2019-03-26 15:07:30,892 platform:- - atomic_reactor.plugin - DEBUG - plugin 'pull_base_image' finished in 0s
2019-03-26 15:07:30,892 platform:- - atomic_reactor.plugin - DEBUG - running plugin 'bump_release'
2019-03-26 15:07:30,892 platform:- - atomic_reactor.plugin - INFO - running plugin instance with args: '{}'
2019-03-26 15:07:30,892 platform:- - atomic_reactor.koji_util - INFO - Using SSL certificates for Koji authentication
2019-03-26 15:07:31,097 platform:- - atomic_reactor.util - DEBUG - query_registry: querying /v2/mbasti-osbs-box/fedora/manifests/latest, headers: {u'Accept': 'application/vnd.docker.distribution.manifest.v1+json'}
2019-03-26 15:07:31,334 platform:- - atomic_reactor.util - DEBUG - query_registry: [401] https://quay.io/v2/mbasti-osbs-box/fedora/manifests/latest
2019-03-26 15:07:31,335 platform:- - atomic_reactor.util - DEBUG - query_registry: [401] https://quay.io/v2/mbasti-osbs-box/fedora/manifests/latest
2019-03-26 15:07:31,335 platform:- - atomic_reactor.util - DEBUG - query_registry: response headers: {'content-length': '112', 'server': 'nginx/1.14.2', 'connection': 'keep-alive', 'docker-distribution-api-version': 'registry/2.0', 'date': 'Tue, 26 Mar 2019 15:07:32 GMT', 'content-type': 'application/json', 'www-authenticate': 'Bearer realm="https://quay.io/v2/auth",service="quay.io",scope="repository:mbasti-osbs-box/fedora:pull"'}
2019-03-26 15:07:31,336 platform:- - atomic_reactor.plugin - DEBUG - Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 257, in run
    plugin_response = plugin_instance.run()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugins/pre_bump_release.py", line 114, in run
    parser = df_parser(self.workflow.builder.df_path, workflow=self.workflow)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1197, in df_parser
    parent_config = workflow.builder.base_image_inspect[INSPECT_CONFIG]
  File "/usr/lib/python2.7/site-packages/atomic_reactor/build.py", line 288, in base_image_inspect
    self.base_image_insecure)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1056, in get_inspect_for_image
    dockercfg_path=dockercfg_path)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1037, in get_all_manifests
    response, _ = get_manifest(image, registry_session, version)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 916, in get_manifest
    response = query_registry(registry_session, image, digest=None, version=version)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 853, in query_registry
    response.raise_for_status()
  File "/usr/lib/python2.7/site-packages/requests/models.py", line 834, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 401 Client Error: UNAUTHORIZED
```

mlangsdorf commented:
Updated to handle the pre_bump_release case.

MartinBasti commented on atomic_reactor/plugins/pre_pull_base_image.py:
```diff
@@ -302,15 +305,18 @@ def _get_manifest_list(self, image):
         if image in self.manifest_list_cache:
             return self.manifest_list_cache[image]
 
+        self.log.debug("\tMJLL - get manifest for image %s from registry %s which is %s secure and has secret config %s", image, image.registry, self.parent_registry_insecure, self.parent_registry_secret)
```
Just marking this to get rid of that secret in final version

MartinBasti commented:
We are going forward, this one is from worker:

```
2019-03-28 15:24:45,658 - atomic_reactor.util - ERROR - {u'errorDetail': {u'message': u'Error: Status 403 trying to pull repository mbasti-osbs-box/fedora: "{\\"error\\": \\"Permission Denied\\"}"', u'code': 403}, u'error': u'Error: Status 403 trying to pull repository mbasti-osbs-box/fedora: "{\\"error\\": \\"Permission Denied\\"}"'}
2019-03-28 15:24:45,658 - atomic_reactor.util - INFO - no more logs
2019-03-28 15:24:45,659 - atomic_reactor.plugin - DEBUG - Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 257, in run
    plugin_response = plugin_instance.run()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugins/pre_pull_base_image.py", line 119, in run
    new_image = self._pull_and_tag_image(image, build_json, str(nonce))
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugins/pre_pull_base_image.py", line 258, in _pull_and_tag_image
    self.tasker.pull_image(image, insecure=self.parent_registry_insecure)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/core.py", line 531, in pull_image
    decode=True, stream=True)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/core.py", line 332, in retry_generator
    exc)
RetryGeneratorException: Failed to pull image (u'quay.io/mbasti-osbs-box/fedora',): u'Error: Status 403 trying to pull repository mbasti-osbs-box/fedora: "{\\"error\\": \\"Permission Denied\\"}"'
```

it seems that `tasker` misses secret configuration

twaugh commented on atomic_reactor/build.py:
```diff
@@ -246,7 +246,7 @@ def recreate_parent_images(self):
         self.parent_images = parent_images
 
-    def set_base_image(self, base_image, parents_pulled=True, insecure=False):
+    def set_base_image(self, base_image, parents_pulled=True, insecure=False, secret=None):
```
Maybe we could call this parameter `dockercfg_path`? It's more descriptive.

twaugh commented on atomic_reactor/plugins/pre_reactor_config.py:
```diff
@@ -278,10 +278,14 @@ def get_source_registry(workflow, fallback=NO_FALLBACK):
         'uri': RegistryURI(source_registry['url']),
-        'insecure': source_registry.get('insecure', False)
+        'insecure': source_registry.get('insecure', False),
+        'secret': None
```
Could we call this `dockercfg_path` to make it clearer what value to expect there?

mlangsdorf replied:
Yes

twaugh commented:
Please include the JIRA issue name in the commit message.

twaugh commented on atomic_reactor/plugins/pre_pull_base_image.py:
```diff
@@ -303,14 +307,16 @@ def _get_manifest_list(self, image):
                 config_blob = get_config_from_registry(image, image.registry, image.tag,
-                                                       insecure=self.parent_registry_insecure)
+                                                       insecure=self.parent_registry_insecure,
+                                                       dockercfg_path=self.parent_registry_dockercfg_path)
```
One way around this would be to format it like:
```
config_blob = get_config_from_registry(
    image, ...
    ...
)
```
We've done that in other places.

mlangsdorf replied:
Fixed.

mlangsdorf commented:
Seems to be working in my tests, stickler is satisfied.

twaugh commented on atomic_reactor/core.py:
```diff
@@ -517,17 +517,21 @@ def get_image_info_by_image_name(self, image, exact_tag=True):
         tag = image.tag
+
+        if dockercfg_path:
+            self.login(registry=image.registry, docker_secret_path=dockercfg_path)
```
Looks like this line isn't getting unit test coverage.

mlangsdorf replied:
Fixed.

MartinBasti commented:
I tested this and it seems working, however quay.io still doesn't support schema2 for all organizations, so I'm getting: `Missing platforms for registry quay.io: found [], expected ['x86_64']` caused by supporting only v1 schema

mlangsdorf commented:
Added to release notes under improvements:
- Added support for quay.io private repositories for source image pulls.  Some quay.io organizations do not support v2 schema and may generate `Missing platforms for registry quay.io` errors.
